test/unit/org/apache/cassandra/io/tries/WalkerTest.java

```diff
@@ -0,0 +1,231 @@
+        logger.info("Trie dump: \n{}", new String(dumpBuf.getData()));
+        logger.info("Trie toString: {}", it);
+
+        it.goMax(rootPos);
+        assertEquals(7, it.payloadFlags());
+        assertEquals(TrieNode.Types.PAYLOAD_ONLY.ordinal, it.nodeTypeOrdinal());
+        assertEquals(1, it.nodeSize());
+        assertFalse(it.hasChildren());
+
+        it.goMin(rootPos);
+        assertEquals(1, it.payloadFlags());
+        assertEquals(TrieNode.Types.PAYLOAD_ONLY.ordinal, it.nodeTypeOrdinal());
+        assertEquals(1, it.nodeSize());
+        assertFalse(it.hasChildren());
+
+        assertEquals(-1, it.follow(source("151")));
+        assertEquals(2, it.payloadFlags());
+
+        assertEquals('3', it.follow(source("135")));
+
+        assertEquals('3', it.followWithGreater(source("135")));
+        it.goMin(it.greaterBranch);
+        assertEquals(2, it.payloadFlags());
+
+        assertEquals('3', it.followWithLesser(source("135")));
+        it.goMax(it.lesserBranch);
+        assertEquals(1, it.payloadFlags());
+
+        assertEquals(3, (Object) it.prefix(source("155"), (walker, payloadPosition, payloadFlags) -> payloadFlags));
+        assertNull(it.prefix(source("516"), (walker, payloadPosition, payloadFlags) -> payloadFlags));
+        assertEquals(5, (Object) it.prefix(source("5151"), (walker, payloadPosition, payloadFlags) -> payloadFlags));
+        assertEquals(1, (Object) it.prefix(source("1151"), (walker, payloadPosition, payloadFlags) -> payloadFlags));
+
+        assertEquals(3, (Object) it.prefixAndNeighbours(source("155"), (walker, payloadPosition, payloadFlags) -> payloadFlags));
+        assertNull(it.prefixAndNeighbours(source("516"), (walker, payloadPosition, payloadFlags) -> payloadFlags));
+        assertEquals(5, (Object) it.prefixAndNeighbours(source("5151"), (walker, payloadPosition, payloadFlags) -> payloadFlags));
+        assertEquals(1, (Object) it.prefixAndNeighbours(source("1151"), (walker, payloadPosition, payloadFlags) -> payloadFlags));
+
+        assertEquals(3, (Object) it.prefixAndNeighbours(source("1555"), (walker, payloadPosition, payloadFlags) -> payloadFlags));
+        it.goMax(it.lesserBranch);
+        assertEquals(2, it.payloadFlags());
+        it.goMin(it.greaterBranch);
+        assertEquals(4, it.payloadFlags());
+    }
+
+    @Test
+    public void testWithBounds() throws IOException
+    {
+        DataOutputBuffer buf = new DataOutputBufferPaged();
+        IncrementalTrieWriter<Integer> builder = makeTrie(buf);
+        long rootPos = builder.complete();
+
+        Rebufferer source = new ByteBufRebufferer(buf.asNewBuffer());
+
+        InternalIterator it = new InternalIterator(source, rootPos, source("151"), source("515"), false);
+        long pos;
+        assertNotEquals(-1, pos = it.nextPayloadedNode());
+        assertEquals(3, TrieNode.at(buf.asNewBuffer(), (int) pos).payloadFlags(buf.asNewBuffer(), (int) pos));
```

Comment:
This was a little confusing to me at first. "15" is clearly a prefix, but "151" is an exact match for an entry, right? The semantics of "admitting a prefix" in that case just looks like an "inclusive lower bound". In `testWithBoundsAndAdmitPrefix()`, we would also have a passing test with the left bound "15", which seems more like "admitting a prefix", but we start w/ "151" there as well.

Am I thinking about this wrongly? :D

Reply:
Not at all, but there is more to it. We must also return "151" for a "15151" query with `admitPrefix`.

I added some clarifying JavaDoc and more tests, and corrected some errors when `admitPrefix == false`, which we never actually used.